src/main/scala/com/actian/spark_vector/vector/PredicatePushdown.scala

```diff
@@ -0,0 +1,163 @@
+   *     @note: We also include here ("==", x) as a shorter notation for ("[]", x, x)
+   */
+  type ValueRange = Seq[String]
+
+  /**
+   * A `ValueRanges` object is a sequence of `ValueRange` elements and it is associated with a certain column.
+   * As such, all string bounds of all the `ValueRange`s are expected to be literals of that column's `VectorDataType`.
+   */
+  type ValueRanges = Seq[ValueRange]
+
+  /** Creates a `Column` that represents the condition ("predicate"), in Spark's terms, that is expressed by the given `range` */
+  def makeFilterForColumn(column: Column, litToConstColFcn: String => Column, range: ValueRange): Column = {
+    range match {
+      case Seq(cmpType, arg) => {
+        val value = litToConstColFcn(arg)
+        val cond = cmpType match {
+          case "<" => column < value
+          case "<=" => column <= value
+          case "==" => column === value
+          case ">=" => column >= value
+          case ">" => column > value
+        }
+        value.isNull or cond
+      }
+      case Seq(intervalType, arg1, arg2) => {
+        val lowerBound = litToConstColFcn(arg1)
+        val upperBound = litToConstColFcn(arg2)
+        val cond = intervalType match {
+          case "()" => column > lowerBound and column < upperBound
+          case "(]" => column > lowerBound and column <= upperBound
+          case "[)" => column >= lowerBound and column < upperBound
+          case "[]" => column >= lowerBound and column <= upperBound
+        }
+        lowerBound.isNull or upperBound.isNull or cond
+      }
+    }
+  }
+
+  /** Creates a `Column` that represents the condition ("predicate"), in Spark's terms, that is expressed by the given `ranges` */
+  def makeFilterForColumn(column: Column, vectorDataType: VectorDataType.EnumVal, sparkDataType: DataType, ranges: ValueRanges,
+    sparkContext: SparkContext): Column = {
+
+    def litToConstColFcn(vectorLiteral: String): Column = {
+      val sparkLiteral = vectorLiteralToSparkLiteral(vectorLiteral, vectorDataType, sparkContext)
+      lit(sparkLiteral).cast(sparkDataType)
+    }
+
+    ranges match {
+      case Nil => {
+        /* An empty range list means that no tuples should be selected (e.g. col < 0 and col > 0). */
+        expr("false")
+      }
+      case someRanges if typeIsSafeForPredPD(vectorDataType) => {
+        /* The final condition for this column is a disjunction of conditions corresponding to
+         * the individual `ValueRange`s in the list. */
+        someRanges.map(makeFilterForColumn(column, litToConstColFcn, _)).reduceLeft(_ or _)
+      }
+      case default => {
+        logDebug("Predicate pushdown not yet supported for columns of type: " + vectorDataType)
+        expr("true")
+      }
+    }
+  }
+
+  /** Filters the given `df` according to the specified `ValueRanges`. */
+  def applyFiltersToDF(df: DataFrame, columns: Seq[ColumnMetadata], valueRangesMap: Map[String, ValueRanges],
```

Comment:
Since you added the value ranges to ColumnInfo, doesn't it make more sense to add the valueRanges to ColumnMetadata too?